src/Npgsql/TypeHandlers/CompositeHandlers/CompositeHandler.cs

```diff
@@ -152,13 +223,21 @@ static CompositeMemberHandler<T>[] CreateMemberHandlers(PostgresCompositeType pg
                 }
             }
 
-            if (clrHandlerCount == pgFields.Count)
-                return clrHandlers;
+            if (clrMemberHandlerCount != pgFields.Count)
+            {
+                var notMappedFields = string.Join(", ", clrMemberHandlers
+                    .Select((member, memberIndex) => member == null ? $"'{pgFields[memberIndex].Name}'" : null)
+                    .Where(member => member != null));
+                throw new InvalidOperationException($"PostgreSQL composite type {pgType.DisplayName} contains fields {notMappedFields} which could not match any on CLR type {clrType.Name}");
+            }
 
-            var notMappedFields = string.Join(", ", clrHandlers
-                .Select((member, memberIndex) => member == null ? $"'{pgFields[memberIndex].Name}'" : null)
-                .Where(member => member != null));
-            throw new InvalidOperationException($"PostgreSQL composite type {pgType.DisplayName} contains fields {notMappedFields} which could not match any on CLR type {clrType.Name}");
+            _constructor = clrConstructorHandler is null
```

Comment:
So if there is both a parameterless ctor and a matching non-parameterless ctor, we remember both?

Reply:
We remember both, but the parameterless one only to check it existence if there is no other.

Reply:
It's probably safer not to set a variable if you don't intend to use it...

Reply:
Maybe, but there is a probability that in .NET 5 will be a way to call constructors without emitting IL. To specify which constructor should be called you still should know `ConstructorInfo`. So keeping this is for future improvements.

Reply:
Member handlers are required to send some composite back to the backend. Without them it won't be possible (:

Reply:
Ah yes, makes sense :)